### ISSUE TITLE
Argon2 bindings

### ISSUE THREAD

**Issue by dabaer** [potatosalad/erlang-libsodium#2]
This seems to be the only library that is remotely active for libsodium, so I have to put out the word for Argon2, since every major language has support for it except erlang!

https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html

Argon2 is at the top of our minds lately, and we really need native bindings. Because C is scary.

Regards,
Jadenn

**Comment by potatosalad** [comment]
@dabaer I agree that C is scary.

In spite of this, the C waters have been braved and Argon2 support has been added.

Here is example usage translated to Erlang from the [libsodium docs for `crypto_pwhash`](https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html).

```erlang
%% Example 1: key derivation
Password = <<"Correct Horse Battery Staple">>,
KeyLen = libsodium_crypto_box:seedbytes(), % 32
Salt = libsodium_randombytes:buf(libsodium_crypto_pwhash:saltbytes()),% <<204,87,190,254,120,189,9,23,19,113,183,127,62,229,157,201>>
OpslimitInteractive = libsodium_crypto_pwhash:opslimit_interactive(), % 4
MemlimitInteractive = libsodium_crypto_pwhash:memlimit_interactive(), % 33554432
AlgDefault = libsodium_crypto_pwhash:alg_default(). % 1

Key = libsodium_crypto_pwhash:crypto_pwhash(KeyLen, Password, Salt, OpslimitInteractive, MemlimitInteractive, AlgDefault).
% <<91,218,122,209,160,120,58,67,182,252,126,109,208,169,182,156,144,129,60,254,221,71,121,57,112,110,242,78,122,186,62,81>>

%% Example 2: password storage
Password = <<"Correct Horse Battery Staple">>,
OpslimitSensitive = libsodium_crypto_pwhash:opslimit_sensitive(), % 8
MemlimitSensitive = libsodium_crypto_pwhash:memlimit_sensitive(). % 536870912

HashedPassword = libsodium_crypto_pwhash:str(Password, OpslimitSensitive, MemlimitSensitive).
% <<"$argon2i$v=19$m=524288,t=8,p=1$H7OjSWL7hXWOo2EH+Wpn5Q$Ubjt2h+Cca2+YjpzNuy9g3U+tjJTeOEo0sR9hnmIKZQ",0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0>>

libsodium_crypto_pwhash:str_verify(HashedPassword, Password) =:= 0.
% true
```

There are two new modules: `libsodium_crypto_pwhash` and `libsodium_crypto_pwhash_argon2i`.  Using `libsodium_crypto_pwhash_argon2i` should be nearly identical to the examples provided above if you want to ensure Argon2 is being used instead of whatever the default is set to in the future for `libsodium_crypto_pwhash`.

The newly released [libsodium version 0.0.5](https://hex.pm/packages/libsodium/0.0.5) is the one with Argon2 support, so let me know if it works for you or gives you any trouble.

**Comment by potatosalad** [comment]
@dabaer I added some of the test vectors for argon2 from libsodium and discovered some inconsistencies in the values returned by the port driver.  Those inconsistencies have been fixed in [libsodium version 0.0.6](https://hex.pm/packages/libsodium/0.0.6).

Here is an updated example:

```erlang
%% Example 1: key derivation
Password = <<"Correct Horse Battery Staple">>,
KeyLen = libsodium_crypto_box:seedbytes(), % 32
Salt = libsodium_randombytes:buf(libsodium_crypto_pwhash:saltbytes()), % <<204,87,190,254,120,189,9,23,19,113,183,127,62,229,157,201>>
OpslimitInteractive = libsodium_crypto_pwhash:opslimit_interactive(), % 4
MemlimitInteractive = libsodium_crypto_pwhash:memlimit_interactive(), % 33554432
AlgDefault = libsodium_crypto_pwhash:alg_default(). % 1

Key = libsodium_crypto_pwhash:crypto_pwhash(KeyLen, Password, Salt, OpslimitInteractive, MemlimitInteractive, AlgDefault).
% <<91,218,122,209,160,120,58,67,182,252,126,109,208,169,182,156,144,129,60,254,221,71,121,57,112,110,242,78,122,186,62,81>>

%% Example 2: password storage
Password = <<"Correct Horse Battery Staple">>,
OpslimitSensitive = libsodium_crypto_pwhash:opslimit_sensitive(), % 8
MemlimitSensitive = libsodium_crypto_pwhash:memlimit_sensitive(). % 536870912

HashedPassword = libsodium_crypto_pwhash:str(Password, OpslimitSensitive, MemlimitSensitive).
% <<"$argon2i$v=19$m=524288,t=8,p=1$NfCkQpBRhsz0u3N/fItPeg$s57hvAyx4V3e12qV1MROfJm+X0cVpe3jQMKhEryXXLc">>

libsodium_crypto_pwhash:str_verify(HashedPassword, Password) =:= 0.
% true
```

**Comment by dabaer** [comment]
I'm glad someone is brave enough around here :P

*Spams stars*

Thanks!
